### ISSUE TITLE
(car (quote (10 20 30))) should be evaluated to 10

### ISSUE THREAD

**Issue by kariya-mitsuru** [mattn/cisp#12]
**TL;DR**

In the current implementation, `(car '(10 20 30))` is evaluated to `10` but `(car (quote (10 20 30)))` causes `argument is not a list` error, because a node-type of `(10 20 30)` is `NODE_CALL` in the latter case.

I think that there are three options to correct this problem.
1. Change `do_car` to accept `NODE_CALL` type.
2. Abolish `NODE_CALL` type and use `NODE_LIST` type instead.
3. Use `NODE_LIST` type if a first element of the outer list has a `quote` identifier.

First option looks easy but all the other functions may have to be changed too, and the changed functions may look noisy.
Second option looks harder but the changed functions may look simple.
Third option looks too ad-hoc.

IMHO, 2nd option is better than others because `(car (quote (10 20 30)))` is itself a list (so, this language have a name 'LISt-Processing'), and, even if the other option is chosen, you will have to implement `NODE_LIST` evaluation function to support `eval` function (`(eval '(+ 10 20))`).

**Comment by mattn** [comment]
In my opinion, I prefer 1st. Because sbcl handle `(1 2)` as lambda.

![](http://go-gyazo.appspot.com/7529f2fdc28e96b0.png)

In my poor understanding of spec of lisp, the car of the list `(x y)` should be always function call. But it's not strong opinion.

**Comment by mattn** [comment]
hmm, sbcl says

```
(type-of (quote (10 20 30)))
CONS
```

```
(type-of (10 20 30))
CONS
```

```
(type-of (a 20 30))
((LAMBDA ()))
```

AFAICS, seems that sbcl handle that the argument is fuction or list with whether the prefix is non-digits.

**Comment by kariya-mitsuru** [comment]
According to my understanding, LISP processes `(car (quote (1 2)))` like below.
1. Read the string `(car (quote (1 2)))` and construct an ast.
   The ast consists of six cons cells and eight atoms(two symbols, two integers and four `nil`s).  
   That is, `(car . ((quote . ((1 . (2 . nil)) . nil)) . nil) . nil)`.  
   Note that `nil` is really `NULL`.
2. Evaluate the ast.
   If the ast is ...
   1. a symbol, then
      a result is a value that bound to the symbol, then terminate evaluation.  
      Note that lisp-1 (like Scheme) only have one binding value, on the other hands, lisp-2
      (like Common Lisp and cisp's current implementation) have two separate binding values
      (normal value & function value).  
      In lisp-2, it is context-dependent which value is used.  
      In this example above, `car` and `quote` are in this category, so they are evaluated to
      their values, i.e. 'a built-in function car' and 'a special form quote'.
   2. an atom other than symbol, then
      a result is an atom itself, then terminate evaluation.  
      In this example above, `1`, `2` and `nil` are in this category.  
      Note that built-in function and lambda (and even special form) may be considered as
      this category.
   3. a list (i.e. list of cons cells that a cdr part of the last cons cell is nil), then
      go to step 3.
   4. a cons cell but it is not part of a list, then  
      raise an error, then terminate evaluation.
3. Evaluate the list.
   First, evaluate the first element of the list. It is done by recursive evaluation.  
   If the result of the evaluation is ...
   1. built-in function or lambda, then
      evaluate all the rest of the list and apply the function (or lambda) to the result of evaluations.  
      Of course, the evaluation of all the rest of the list is done by recursive evaluations.
   2. special form, then
      apply the special form to all the rest of list.  
      Each element of the list are evaluated if necessary.  
      In this example above, `a special form quote` does not evaluate its argument.  
      On the other hand, `a special form if` evaluates its first argument (i.e. second element of the list)
      always, but evaluates either its second or third argument only, not both.
   3. other than above, then
      raise an error.

Note that it omits macro expansion.  It is done in first of step 3.

**Comment by kariya-mitsuru** [comment]
The sbcl is the implementation of common lisp, so `(type-of (1 2))` is evaluated ...
1.  `(type-of (1 2))` is a list (first element is a symbol `type-of`, second element is
   a list `(1 2)`).  
   So evaluation of it causes evaluation of `type-of`.  
   Because the symbol `type-of` is bound to 'a built-in function type-of', the result
   of evaluation is 'a built-in function type-of.'  
   Since the result is built-in function, it causes evaluation of all rest of the
   list `(1 2)`.  
2.  `(1 2)` is a list (first element is an integer `1`, second element is an integer `2`).  
   So evaluation of it causes evaluation of `1`.  
   Of course, the result of evaluation is `1` itself.  
3.  An evaluation of `(1 2)` **raises an error** since `1` is neither built-in function nor lambda.  

Similarly, `(type-of (a 20 30))` is evaluated ..
1.  `(type-of (a 20 30))` is a list (first element is a symbol `type-of`, second element is
   a list `(a 20 30)`).  
   So evaluation of it causes evaluation of `type-of`.  
   Because the symbol `type-of` is bound to 'a built-in function type-of', the result
   of evaluation is 'a built-in function type-of.'  
   Since the result is built-in function, it causes evaluation of all rest of the
   list `(a 20 30)`.  
2.  `(a 20 30)` is a list (first element is a symbol `a`, second element is an integer `20`, third element is an integer `30`).  
   So evaluation of it causes evaluation of `a`.  
3.  An evaluation of `a` **raises an error** if the symbol `a` is not bound to any function value.

Anyway, it seems that the sbcl handles `(1 2)` and `(a 20 30)` both simple lists.

**Comment by mattn** [comment]
Thank for your explaining. I should consider the implementation and how to fix this issue.

**Comment by kariya-mitsuru** [comment]
I forgot one more thing.

If you plan to implement cons cell modification functions (like `rplaca` and `rplacd` of common lisp, or `set-car!` and `set-cdr!` of scheme), the current `NODE` structure may be hard to support it.

Ex.

``` lisp
(setq x '(1 2 3 4 5))
-> (1 2 3 4 5)
(setq y (cdr x))
-> (2 3 4 5)
(rplaca y 10)
-> (10 3 4 5)
x
-> (1 10 3 4 5)
```

**Comment by mattn** [comment]
I'm currently working this on fix-list branch. This is removed NODE_CALL. But still work in progress.

**Comment by mattn** [comment]
Hmm, rplaca is possible to implement but rplacd resizes original cells like below.

``` lisp
(setq x '(1 2 3 4 5))
(print x)
(setq y (cdr x))
(print y)
(print (rplacd y 10))
(print x)
```

```
(1 2 3 4 5)
(2 3 4 5)
(2 . 10)
(1 2 . 10)
```

**Comment by mattn** [comment]
fixed implementation of data structure. So most of issue you mentioned will be solved. But not perfect.
